hybrid/pkg/server/server_interceptor.go

```diff
@@ -0,0 +1,119 @@
+package hybrid_server
+
+import (
+	"context"
+
+	hclog "github.com/hashicorp/go-hclog"
+	nodeattestorv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/plugin/server/nodeattestor/v1"
+)
+
+type ServerInterceptorInterface interface {
+	Recv() (*nodeattestorv1.AttestRequest, error)
+	Send(resp *nodeattestorv1.AttestResponse) error
+	setCustomStream(stream nodeattestorv1.NodeAttestor_AttestServer)
+	SetContext(ctx context.Context)
+	Context() context.Context
+	SetLogger(logger hclog.Logger)
+	SetReq(req *nodeattestorv1.AttestRequest)
+	CanReattest() []bool
+	SpiffeID() string
+	SetSpiffeID(spiffeID string)
+	CombinedSelectors() []string
+	Stream() nodeattestorv1.NodeAttestor_AttestServer
+	ResetInterceptor()
+	SpawnInterceptor() ServerInterceptorInterface
+}
+
+type HybridPluginServerInterceptor struct {
+	ctx    context.Context
+	stream nodeattestorv1.NodeAttestor_AttestServer
+	nodeattestorv1.NodeAttestor_AttestServer
+	logger            hclog.Logger
+	req               *nodeattestorv1.AttestRequest
+	Response          *nodeattestorv1.AttestResponse
+	combinedSelectors []string
+	spiffeID          string
+	canReattest       []bool
+}
+
+func (m *HybridPluginServerInterceptor) ResetInterceptor() {
+	m.ctx = nil
+	m.stream = nil
+	m.logger = nil
+	m.req = nil
+	m.Response = nil
+	m.combinedSelectors = nil
+	m.spiffeID = ""
+	m.canReattest = nil
+}
+
+func (m *HybridPluginServerInterceptor) SpawnInterceptor() ServerInterceptorInterface {
+	return &HybridPluginServerInterceptor{
+		ctx:               m.ctx,
+		stream:            m.stream,
+		logger:            m.logger,
+		req:               m.req,
+		Response:          m.Response,
+		combinedSelectors: m.combinedSelectors,
+		spiffeID:          m.spiffeID,
+		canReattest:       m.canReattest,
+	}
+}
+
+func (m *HybridPluginServerInterceptor) Recv() (*nodeattestorv1.AttestRequest, error) {
+	return m.req, nil // add error here
+}
+
+func (m *HybridPluginServerInterceptor) setCustomStream(stream nodeattestorv1.NodeAttestor_AttestServer) {
+	m.stream = stream
+}
+
+func (m *HybridPluginServerInterceptor) Send(resp *nodeattestorv1.AttestResponse) error {
+	if x, ok := resp.GetResponse().(*nodeattestorv1.AttestResponse_AgentAttributes); ok {
+
+		m.combinedSelectors = append(m.combinedSelectors, x.AgentAttributes.SelectorValues...)
+		if len(m.spiffeID) == 0 {
```

Comment:
this one only captures the spiffeID if no other has been captured before. There are no calls made to datastore. It's just a security to prevent overriding of spiffeIDs

Reply:
Oh no this has nothing do to with data store :) this checks if the parameter passed back from the attestor response, called canReattest is true or false. Which means that if we have a false inside the array values, it means that one of the plugins are not allowed to reattest and in that case, we will check the store using the spiffeID of the first plugin used on the hybrid.

Reply:
but is it not calling `agentstore.IsAttested`? [link](https://github.com/HewlettPackard/roven/pull/3/files#diff-31953d20421891c452a75a1e54e87268b41d79ae62cee2f132448c3d46829545R52-R54) and that one is the one taht is calling datastore